File: app/src/main/java/com/benoitquenaudon/tvfoot/red/app/domain/matches/state/MatchesResult.java
package com.benoitquenaudon.tvfoot.red.app.domain.matches.state;

import com.benoitquenaudon.tvfoot.red.app.data.entity.Match;
import com.google.auto.value.AutoValue;
import java.util.List;
import javax.annotation.Nullable;

import static com.benoitquenaudon.tvfoot.red.app.domain.matches.state.MatchesResult.Status.NEXT_PAGE_FAILURE;
import static com.benoitquenaudon.tvfoot.red.app.domain.matches.state.MatchesResult.Status.NEXT_PAGE_IN_FLIGHT;
import static com.benoitquenaudon.tvfoot.red.app.domain.matches.state.MatchesResult.Status.NEXT_PAGE_SUCCESS;
import static com.benoitquenaudon.tvfoot.red.app.domain.matches.state.MatchesResult.Status.REFRESH_FAILURE;
import static com.benoitquenaudon.tvfoot.red.app.domain.matches.state.MatchesResult.Status.REFRESH_IN_FLIGHT;
import static com.benoitquenaudon.tvfoot.red.app.domain.matches.state.MatchesResult.Status.REFRESH_SUCCESS;

interface MatchesResult {
  @AutoValue abstract class RefreshResult implements MatchesResult {
    abstract Status status();

    @Nullable abstract List<Match> matches();

    @Nullable abstract Throwable throwable();

    static RefreshResult success(List<Match> matches) {
      return new AutoValue_MatchesResult_RefreshResult(REFRESH_SUCCESS, matches, null);
    }

    static RefreshResult failure(Throwable throwable) {
      return new AutoValue_MatchesResult_RefreshResult(REFRESH_FAILURE, null, throwable);
    }

    static RefreshResult inFlight() {
      return new AutoValue_MatchesResult_RefreshResult(REFRESH_IN_FLIGHT, null, null);
    }
  }

  @AutoValue abstract class GetLastStateResult implements MatchesResult {
    static GetLastStateResult create() {
      return new AutoValue_MatchesResult_GetLastStateResult();
    }
  }

  @AutoValue abstract class LoadNextPageResult implements MatchesResult {
    abstract Status status();

    @Nullable abstract List<Match> matches();

    @Nullable abstract Throwable error();

    abstract int pageIndex();

    static LoadNextPageResult success(int pageIndex, List<Match> matches) {
      return new AutoValue_MatchesResult_LoadNextPageResult(NEXT_PAGE_SUCCESS, matches, null,
          pageIndex);
    }

    static LoadNextPageResult failure(Throwable throwable) {
      return new AutoValue_MatchesResult_LoadNextPageResult(NEXT_PAGE_FAILURE, null, throwable, -1);
    }

    static LoadNextPageResult inFlight() {
      return new AutoValue_MatchesResult_LoadNextPageResult(NEXT_PAGE_IN_FLIGHT, null, null, -1);
    }
  }

  enum Status {
    REFRESH_IN_FLIGHT, REFRESH_FAILURE, REFRESH_SUCCESS, //
    NEXT_PAGE_IN_FLIGHT, NEXT_PAGE_FAILURE, NEXT_PAGE_SUCCESS, //
    PULL_TO_REFRESH_IN_FLIGHT, PULL_TO_REFRESH_FAILURE, PULL_TO_REFRESH_SUCCESS, //
  }
}

File: app/src/main/java/com/benoitquenaudon/tvfoot/red/app/domain/matches/state/MatchesResult.kt
package com.benoitquenaudon.tvfoot.red.app.domain.matches.state

import com.benoitquenaudon.tvfoot.red.app.common.LceStatus
import com.benoitquenaudon.tvfoot.red.app.common.LceStatus.FAILURE
import com.benoitquenaudon.tvfoot.red.app.common.LceStatus.IN_FLIGHT
import com.benoitquenaudon.tvfoot.red.app.common.LceStatus.SUCCESS
import com.benoitquenaudon.tvfoot.red.app.data.entity.Match
import com.benoitquenaudon.tvfoot.red.app.data.entity.Tag
import com.benoitquenaudon.tvfoot.red.app.mvi.MviResult

sealed class MatchesResult : MviResult {
  @Suppress("DataClassPrivateConstructor")
  data class RefreshResult private constructor(
      val status: LceStatus,
      val matches: List<Match>?,
      val throwable: Throwable?
  ) : MatchesResult() {
    companion object Factory {
      fun success(matches: List<Match>): RefreshResult {
        return RefreshResult(SUCCESS, matches, null)
      }

      fun failure(throwable: Throwable): RefreshResult {
        return RefreshResult(FAILURE, null, throwable)
      }

      fun inFlight(): RefreshResult {
        return RefreshResult(IN_FLIGHT, null, null)
      }
    }
  }

  object GetLastStateResult : MatchesResult()

  @Suppress("DataClassPrivateConstructor")
  data class LoadNextPageResult private constructor(
      val status: LceStatus,
      val matches: List<Match>?,
      val error: Throwable?,
      val pageIndex: Int
  ) : MatchesResult() {
    companion object Factory {
      fun success(pageIndex: Int, matches: List<Match>): LoadNextPageResult {
        return LoadNextPageResult(SUCCESS, matches, null, pageIndex)
      }

      fun failure(throwable: Throwable): LoadNextPageResult {
        return LoadNextPageResult(FAILURE, null, throwable, -1)
      }

      fun inFlight(): LoadNextPageResult {
        return LoadNextPageResult(IN_FLIGHT, null, null, -1)
      }
    }
  }

  object ClearFiltersResult : MatchesResult()

  data class ToggleFilterResult(val tagName: String) : MatchesResult()

  @Suppress("DataClassPrivateConstructor")
  data class LoadTagsResult private constructor(
      val status: LceStatus,
      val tags: List<Tag>?,
      val error: Throwable?
  ) : MatchesResult() {
    companion object Factory {
      fun success(tags: List<Tag>): LoadTagsResult {
        return LoadTagsResult(SUCCESS, tags, null)
      }

      fun failure(throwable: Throwable): LoadTagsResult {
        return LoadTagsResult(FAILURE, null, throwable)
      }

      fun inFlight(): LoadTagsResult {
        return LoadTagsResult(IN_FLIGHT, null, null)
      }
    }
  }
}
